active/001-tabular-msa.md

```diff
@@ -0,0 +1,292 @@
+
+In scikit-bio 0.4.0 the sequence classes were completely redesigned and the
+current MSA class (`Alignment`) makes invalid assumptions about the API of the
+sequences it stores (for example, see [#1036](https://github.com/biocore/scikit-bio/issues/1036)). This gives us an
+opportunity to rework the MSA API to better fit the new sequence classes.
+
+After these changes, we will have an obvious way to introduce new MSA
+structures (e.g., partial-order MSA), better interaction with the new sequence
+API, pandas-style vectorization, and flexible, arbitrary metadata that can
+represent file formats such as
+[Stockholm](http://sonnhammer.sbc.su.se/Stockholm.html).
+
+# Detailed design
+
+A `TabularMSA` object is a mutable, monomorphic, ordered collection of
+`Sequence` objects that are of equal length. As its name implies, `TabularMSA`
+represents a table-like structure where the first dimension is sequences and
+the second dimension is positions in the alignment. Thus, each row corresponds
+to a sequence and each column corresponds to a position. The first dimension is
+optionally indexed by an arbitrary key. Both dimensions are indexable via
+ordinals.
+
+One of the advantages of `TabularMSA` over the existing `Alignment` class is
+vectorization support and boolean indexing. Currently, to filter out alignment
+positions that have a Shannon entropy greater than 0.5:
+
+```python
+aln.subalignment(positions_to_keep=[
+    i for i, v in enumerate(aln.position_entropies()) if v <= 0.5])
+```
+
+With `TabularMSA`:
+
+```python
+aln[:, aln.entropies() <= 0.5]
+```
+
+Another example is filtering out completely-gapped positions. With the current
+API, this will (probably) work:
+
+```python
+aln.omit_gap_positions(1.0 - np.finfo(float).eps)
+```
+
+With `TabularMSA`:
+
+```python
+aln[:, aln.gap_frequencies(relative=True) < 1.0]
+```
+
+Thus with boolean vector support, `TabularMSA` is significantly more flexible
+in its filtering capabilities.
+
+## Constructors
+- `TabularMSA(iterable, key=None)`
+
+   Construct a `TabularMSA` from an iterable of `Sequence` objects. `key` is a
+   callable or a lookup key in `Sequence.metadata`. When `key=None`, will be
+   the ordinals 0 through N - 1.
+
+## Properties
+- `dtype -> Sequence type`
+
+   Type of sequence stored (e.g., `DNA`)
+- `shape -> (int, int)`
+
+   Tuple storing number of rows (sequences) and columns (positions)
+- `loc -> indexer`
+
+   Allows indexing via keys (like pandas)
+- `iloc -> indexer`
+
+   Allows indexing via indices (like pandas)
+
+- `metadata -> dict`
+
+   Arbitrary metadata that applies to the alignment as a whole.
+
+- `positional_metadata -> pd.DataFrame`
+
+   Arbitrary metadata that applies to the positions of the alignment as a
+   whole.
+
+## Methods
+- `__nonzero__() -> bool`
+
+   Return `True` if MSA has one or more sequences
+- `__len__() -> int`
+
+   Return number of sequences in MSA. This is equivalent to `aln.shape[0]`
```

Comment:
I think it is important for our objects to support as many of the magic methods as possible in order to encourage Pythonic usage of the library. We can provide explicit methods in addition to the magics when necessary (i.e., `aln.shape`).

Since `TabularMSA` is conceptually a list of `Sequence` objects (rows), it makes sense to have `len` return the number of sequences (similar to how calling `len` on a 2D `ndarray` tells you the number of rows). This may be confusing to users initially if they have a transposed mental model, but as soon as they start using the object or read any of the docs it should become clear which axis these operations apply to.